pkg/container/containerd.go

```diff
@@ -166,3 +172,345 @@ func (d *DisableContainerd) Execute(runtime connector.Runtime) error {
+		Hosts: []connector.Host{host},
+		Prepare: &prepare.PrepareCollection{
+			&ContainerdExist{Not: true},
+		},
+		Action: &action.Template{
+			Template: templates.ContainerdConfig,
+			Dst:      filepath.Join("/etc/containerd/", templates.ContainerdConfig.Name()),
+			Data: util.Data{
+				"Mirrors":            templates.Mirrors(kubeAction.KubeConf),
+				"InsecureRegistries": kubeAction.KubeConf.Cluster.Registry.InsecureRegistries,
+				"SandBoxImage":       images.GetImage(runtime, kubeAction.KubeConf, "pause").ImageName(),
+				"Auths":              registry.DockerRegistryAuthEntries(kubeAction.KubeConf.Cluster.Registry.Auths),
+				"DataRoot":           templates.DataRoot(kubeAction.KubeConf),
+			},
+		},
+		Parallel: false,
+	}
+
+	generateCrictlConfig := &task.RemoteTask{
+		Name:  "GenerateCrictlConfig",
+		Desc:  "Generate crictl config",
+		Hosts: []connector.Host{host},
+		Prepare: &prepare.PrepareCollection{
+			&ContainerdExist{Not: true},
+		},
+		Action: &action.Template{
+			Template: templates.CrictlConfig,
+			Dst:      filepath.Join("/etc/", templates.CrictlConfig.Name()),
+			Data: util.Data{
+				"Endpoint": kubeAction.KubeConf.Cluster.Kubernetes.ContainerRuntimeEndpoint,
+			},
+		},
+		Parallel: false,
+	}
+
+	enableContainerd := &task.RemoteTask{
+		Name:  "EnableContainerd",
+		Desc:  "Enable containerd",
+		Hosts: []connector.Host{host},
+		// Prepare: &prepare.PrepareCollection{
+		// 	&ContainerdExist{Not: true},
+		// },
+		Action:   new(EnableContainerd),
+		Parallel: false,
+	}
+	tasks := []task.Interface{
+		syncContainerd,
+		syncCrictlBinaries,
+		generateContainerdService,
+		generateContainerdConfig,
+		generateCrictlConfig,
+		enableContainerd,
+	}
+
+	for i := range tasks {
+		t := tasks[i]
+		t.Init(runtime, kubeAction.ModuleCache, kubeAction.PipelineCache)
+		if res := t.Execute(); res.IsFailed() {
+			return res.CombineErr()
+		}
+	}
+	return nil
+}
+
+type MigrateSelfNodeToDocker struct {
+	common.KubeAction
+}
+
+func (d *MigrateSelfNodeToDocker) Execute(runtime connector.Runtime) error {
+	nodeName, _ := runtime.GetRunner().SudoCmd("hostname", true)
+	if _, err := runtime.GetRunner().SudoCmd(fmt.Sprintf("/usr/local/bin/kubectl cordon %s ", nodeName), true); err != nil {
+		return errors.Wrap(err, fmt.Sprintf("cordon the node: %s failed", nodeName))
+	}
+	if _, err := runtime.GetRunner().SudoCmd(fmt.Sprintf(
+		"/usr/local/bin/kubectl drain %s --delete-emptydir-data --ignore-daemonsets --timeout=2m --force", nodeName),
+		true); err != nil {
+		return errors.Wrap(err, fmt.Sprintf("drain the node: %s failed", nodeName))
+	}
+	// if _, err := runtime.GetRunner().SudoCmd("systemctl stop kubelet", false); err != nil {
+	// 	return errors.Wrap(errors.WithStack(err), fmt.Sprintf("stop kubelet failed: %s", nodeName))
+	// }
+	if err := UninstallContainerTasks(runtime, d.KubeAction); err != nil {
+		return errors.Wrap(errors.WithStack(err), fmt.Sprintf("Uninstall ContainerTasks failed: %s", nodeName))
+	}
```

Comment:
I see these codes are duplicated with `MigrateSelfNodeTocontainerd`, so I suggest modifying them as a new task.
Just an advice, maybe after that the `MigrateSelfNodeToDocker` can become to:
```go
func (d *MigrateSelfNodeToDocker) Execute(runtime connector.Runtime) error {
  CordonTask
  DrainTask
  uninstallTask
  MigrateTask
  RestartTask
  UnCordonTask
}
```

Reply:
thanks, a good advice, I will optimize it,soon